Feature/training api pattern state

Xam1l1on opened the pull request.

Xam1l1on committed:
Создал ветвь.
Почистил библиотеки в Program.cs

Xam1l1on committed:
Создал ветвь.
Почистил библиотеки в Program.cs

Xam1l1on committed:
Создал ветвь.
Почистил библиотеки в Program.cs

Xam1l1on committed:
Добавил паттерн State (состояние).
Пришлось поменять сущность и дто.

Xam1l1on committed:
Добавил паттерн State (состояние).
Пришлось поменять сущность и дто.

Xam1l1on committed:
Добавил паттерн State (состояние).
Пришлось поменять сущность и дто.

Xam1l1on committed:
Добавил паттерн State (состояние).
Пришлось поменять сущность и дто.

Xam1l1on committed:
Исправил ошибку

Xam1l1on committed:
Обновил сервис: добавил реализацию состояния тренировки.
Зависимость состояний от интерфейса ITrainingState.

Xam1l1on committed:
Обновил сервис: добавил реализацию состояния тренировки.
Зависимость состояний от интерфейса ITrainingState.

Xam1l1on committed:
Обновил сервис: добавил реализацию состояния тренировки.
Зависимость состояний от интерфейса ITrainingState.

Xam1l1on committed:
Обновил сервис: добавил реализацию состояния тренировки.
Зависимость состояний от интерфейса ITrainingState.

lasm16 commented on TrainingsApi/BLL/States/CancelledState.cs:
```diff
@@ -0,0 +1,20 @@
+    {
+        public void Start(TrainingContext context)
+        {
+            throw new InvalidOperationException("Cannot start a cancelled training.");
```
по-хорошему должна быть какая-то бизнес-логика, но сойдет и так

lasm16 commented on TrainingsApi/BLL/States/PlannedState.cs:
```diff
@@ -0,0 +1,20 @@
+﻿namespace TrainingsApi.BLL.States
+{
+    public class PlannedState: ITrainingState
```
тут тоже переименовать

lasm16 commented on TrainingsApi/BLL/Services/ITrainingService.cs:
```diff
@@ -7,5 +7,8 @@ public interface ITrainingService
         Task CreateAsync(TrainingDto dto, CancellationToken cancellationToken = default);
         Task UpdateAsync(TrainingDto dto, CancellationToken cancellationToken = default);
         Task DeleteAsync(int id, CancellationToken cancellationToken = default);
+        Task StartTrainingAsync(int id, CancellationToken cancellationToken = default);
```
вынести в отдельный интрефейс

lasm16 commented on TrainingsApi/BLL/Services/TrainingService.cs:
```diff
@@ -48,5 +49,55 @@ public async Task DeleteAsync(int id, CancellationToken cancellationToken = defa
             await repository.DeleteAsync(training, cancellationToken);
         }
+
+        public async Task StartTrainingAsync(int id, CancellationToken cancellationToken = default)
```
кажется, как будто идет нарушение принципа единственной ответственности. после получение id все эти манипуляции со состояниями должен делать какой-то другой класс.

lasm16 commented on TrainingsApi/Controllers/TrainingsController.cs:
```diff
@@ -47,5 +47,25 @@ public async Task<IActionResult> DeleteTrainingAsync([FromRoute] int id)
         }
+        [HttpPost("{id:int}/start")]
+        public async Task<IActionResult> StartTrainingAsync([FromRoute] int id)
+        {
```
Это тоже все уйдет. Все будет через апдейт

lasm16 reviewed (changes requested):
Нужно доработать

Xam1l1on committed:
Исправил ошибки

Xam1l1on committed:
Исправил ошибки

Xam1l1on committed:
Исправил ошибки

Xam1l1on committed:
Исправил ошибки

Xam1l1on committed:
Изменение имен классов

Xam1l1on committed:
Изменение имен классов

Xam1l1on committed:
Изменение имен классов

Xam1l1on committed:
Изменение имен классов

Xam1l1on committed:
-Добавил docker файл и compose в проект.
-Удалил User из Program
-Добавил в TrainingContext статус null, чтобы по умолчанию создавалась тренировка с Planned статусом

Xam1l1on committed:
-Добавил docker файл и compose в проект.
-Удалил User из Program
-Добавил в TrainingContext статус null, чтобы по умолчанию создавалась тренировка с Planned статусом

Xam1l1on committed:
-Добавил docker файл и compose в проект.
-Удалил User из Program
-Добавил в TrainingContext статус null, чтобы по умолчанию создавалась тренировка с Planned статусом

Xam1l1on committed:
Теперь pattern state отрабатывает.

Xam1l1on committed:
Теперь pattern state отрабатывает.

Xam1l1on committed:
Теперь pattern state отрабатывает.

Xam1l1on committed:
Теперь pattern state отрабатывает.

Xam1l1on committed:
Merge branch 'dev' into feature/training_api_pattern_state

lasm16 commented on TrainingsApi/Repositories/TrainingRepository.cs:
```diff
@@ -33,8 +31,6 @@ public async Task UpdateAsync(Training training, CancellationToken cancellationT
 
         public async Task DeleteAsync(Training training, CancellationToken cancellationToken = default)
         {
-            training.IsDeleted = true;
```
Вернуть

lasm16 commented on TrainingsApi/Controllers/TrainingsController.cs:
```diff
@@ -17,34 +17,47 @@ public async Task<IActionResult> GetTrainingAsync([FromRoute] int id)
         public async Task<IActionResult> DeleteTrainingAsync([FromRoute] int id)
         {
-            await trainingService.DeleteAsync(id);
+            var dto = new TrainingDeleteDto
```
оставим id, дто не нужно

lasm16 commented on TrainingsApi/BLL/Services/TrainingStateService.cs:
```diff
@@ -0,0 +1,59 @@
+
+namespace TrainingsApi.BLL.Services
+{
+    public class TrainingStateService(ITrainingRepository repository) : ITrainingStateService
```
Удалить этот класс

lasm16 commented on TrainingsApi/BLL/States/TrainingContext.cs:
public Training Training { get; set; }
заменить на public string Status { get; set; }

Xam1l1on committed:
фыфв

Xam1l1on committed:
фыфв

Xam1l1on committed:
program убрал UserApi

lasm16 committed:
Добавил кеширование данных

lasm16 committed:
добавил тест на кэш

lasm16 committed:
Merge pull request #21 from lasm16/feature/caching_user_api

feature/caching_user_api

lasm16 committed:
Удалил наблюдатель, добавил билдер

lasm16 committed:
Merge pull request #22 from lasm16/feature/builder_pattern_user_api

feature/builder_pattern_user_api

Vitaly0728 committed:
Feature/add health check middleware (#19)

Health_Cheack Service Add
Co-authored-by: lasm16 <lasm16@yandex.ru>

lasm16 committed:
hot fix

lasm16 committed:
сделал рефакторинг

lasm16 committed:
Merge pull request #24 from lasm16/feature/refactoring_achievements_api

feature/refactoring_achievements_api

lasm16 committed:
сделал рефакторинг

lasm16 committed:
Merge pull request #25 from lasm16/feature/refactoring_nutritions_api

feature/refactoring_nutritions_api

lasm16 committed:
сделал рефакторинг

lasm16 committed:
Merge pull request #26 from lasm16/feature/refactoring_users_api

feature/refactoring_users_api

lasm16 committed:
Сделано

lasm16 committed:
123

lasm16 committed:
abrc
